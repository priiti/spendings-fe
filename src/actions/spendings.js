import uuid from 'uuid';
import {
  ADD_SPENDING,
  EDIT_SPENDING,
  REMOVE_SPENDING
} from './../constants/spendingsConstants';

/**
 * @param {string} description Description for spending, default empty
 * @param {string} note Additional description for spending, default empty
 * @param {number} amount Spending amount, default 0
 * @param {number} createdAt Spending create date, default 0
 */
export const addSpending = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
  type: ADD_SPENDING,
  spending: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

/**
 * @param {string} id Specify id to remove a spending, default empty
 * @param {object} updates Specify spending items to be updated
 */
export const editSpending = (id, updates) => ({
  type: EDIT_SPENDING,
  id,
  updates
});

/**
 * @param {string} id Specify id to remove a spending, default empty
 */
export const removeSpending = ({ id = '' } = {}) => ({
  type: REMOVE_SPENDING,
  id
});
