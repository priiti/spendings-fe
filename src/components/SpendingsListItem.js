import React from 'react';

const SpendingsListItem = (props) => {
  const { description, amount, createdAt } = props;

  return (
    <div>
      <h3>{description}</h3>
      <p>{amount}-{createdAt}</p>
    </div>
  );
};

export default SpendingsListItem;
