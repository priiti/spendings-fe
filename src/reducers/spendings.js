import {
  ADD_SPENDING,
  EDIT_SPENDING,
  REMOVE_SPENDING
} from './../constants/spendingsConstants';

const spendingsReducerDefaultState = [];

const spendingsReducer = (state = spendingsReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_SPENDING:
      return [...state, action.spending]

    case REMOVE_SPENDING:
      return state.filter((spending) => spending.id !== action.id);

    case EDIT_SPENDING:
      return state.map((spending) => {
        if (spending.id === action.id)
          return {
            ...spending,
            ...action.updates
          };
        else
          return spending;
      });
      
    default:
      return state;
  }
};

export default spendingsReducer;
