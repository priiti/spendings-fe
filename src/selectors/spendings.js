import moment from 'moment';
/**
 * 
 * @param {any} spendings Current spendings which will be filtered
 * @param {any} filters Filters used on spendings
 * @returns Filtered spendings
 */
const getFilteredSpendings = (spendings, filters) => {
  const { text, sortBy, startDate, endDate } = filters;
  
  return spendings.filter((spending) => {
    const createdAtMoment = moment(spending.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = spending.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date')
      return a.createdAt < b.createdAt ? 1 : -1;
    else if (sortBy === 'amount')
      return a.amount < b.amount ? 1 : -1;
  });
};

export default getFilteredSpendings;
