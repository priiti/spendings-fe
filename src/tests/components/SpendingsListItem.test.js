import React from 'react';
import { shallow } from 'enzyme';
import spendings from './../fixtures/spendings';
import SpendingsListItem from './../../components/SpendingsListItem';

test('should render SpendingsListItem correctly', () => {
  const wrapper = shallow(<SpendingsListItem {...spendings[0]} />);
  expect(wrapper).toMatchSnapshot();
});
