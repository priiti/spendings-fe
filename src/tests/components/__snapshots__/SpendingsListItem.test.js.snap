// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SpendingsListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Lunch
    </h3>
  </Link>
  <p>
    450
    -
    0
  </p>
</div>
`;
